File: Year-2022/python/Day01/day01.py
from myutils import run_test


@run_test("day01_test.txt", 24000)
def day01_part1(day_input: str):
    calories = read_input(day_input)
    max_calories = 0
    for elf in calories:
        supplies = sum(elf)
        if supplies > max_calories:
            max_calories = supplies
    return max_calories


@run_test("day01_test.txt", 45000)
def day01_part2(day_input: str):
    calories = read_input(day_input)
    sorted_calories = sorted([sum(elf) for elf in calories], reverse=True)
    return sum(sorted_calories[:3])


def read_input(file: str):
    calories = []
    with open(file) as f:
        elf = []
        for line in f.readlines():
            if line.strip() == "":
                calories.append(elf)
                elf = []
            else:
                elf.append(int(line))
    calories.append(elf)
    return calories


if __name__ == "__main__":
    print("Part 1:\t", day01_part1("day01.txt"))
    print("Part 2:\t", day01_part2("day01.txt"))

File: Year-2022/perl/Day01/day01.pl
use strict;
use warnings;
use autodie;

use feature 'say';

# Part 1
open(my $file, '<', 'Day01/day01.txt') or die "Couldn't open file!";
my $max_calories = 0;
my $calories = 0;
while (my $line = <$file>) {
    chomp($line);
    if ($line eq "") {
        if ($calories > $max_calories) {
            $max_calories = $calories;
        }
        $calories = 0;
    } else {
        $calories += $line;
    }
}

say $max_calories;
close($file);

# Part 2
open($file, '<', 'Day01/day01.txt') or die "Couldn't open file!";
my @max_elf_arr = (0, 0, 0);
my $elf = 0;
while (my $line = <$file>) {
    chomp($line);
    if ($line eq "") {
        # if larger than lowest number then add it to array,
        # sort it, and pop lowest number
        if ($elf > $max_elf_arr[0]) {
            push(@max_elf_arr, $elf);
            @max_elf_arr = sort {$a <=> $b} @max_elf_arr;
            shift(@max_elf_arr);
        }
        $elf = 0;
    } else {
        $elf += $line;
    }
}
my $sum = 0;
foreach my $e (@max_elf_arr) {
    $sum += $e;
}
say $sum;
close($file);
